File: WishCalculator.py
import numpy as np

def main():
    #单次抽卡的概率
    #角色池
    def percent_character(x):
        if x <=73:
            return 0.006
        elif x <= 89:
            return (0.006+0.06*(x-73))
        else:
            return 1
    #武器池
    def percent_weapon(x):
        if x <=62:
            return 0.007
        elif x <= 73:
            return (0.007+0.07*(x-62))
        elif x <= 79:
            return (0.777+0.035*(x-73))
        else:
            return 1
    #初始化一个零矩阵
    size = 180*ExpectedCharacterNum+400*ExpectedWeaponNum+1 #这里加上1是为了让最后一行表示达成抽卡预期的状态
    TPmatrix = np.zeros((size, size))
    #角色池的初始状态设置
    CharacterPoolOffset = 0
    if ExpectedCharacterNum != 0:
        if CharacterPoolGuarantee == False:
            CharacterPoolOffset = CharacterPoolStage
        elif CharacterPoolGuarantee == True:
            CharacterPoolOffset = CharacterPoolStage+90
    #生成转移概率矩阵（矩阵前面的行是武器，后面的行是角色，最后一行表示的状态是已经达成抽卡预期）
    #这一部分代码生成抽武器的状态，如果要抽的武器数为0，那么就不会运行这一部分代码
    for i in range(0, ExpectedWeaponNum):
        offset = 400*i
        for j in range(0, 80):
            x = j % 80 + 1
            if i == ExpectedWeaponNum-1:
                #该行属于要抽的最后一把武器的部分，那么如果出限定就会进入角色部分，要加上角色池的初始偏移量
                TPmatrix[offset+j, offset+400+CharacterPoolOffset] = percent_weapon(x)*0.375
            else:
                #该行不属于要抽的最后一把武器的部分，那么抽完会进入下一把武器
                TPmatrix[offset+j, offset+400] = percent_weapon(x)*0.375
            TPmatrix[offset+j, offset+160] = percent_weapon(x)*0.375
            TPmatrix[offset+j, offset+240] = percent_weapon(x)*0.25
            TPmatrix[offset+j, offset+j+1] = 1-percent_weapon(x)
        for j in range(80, 160):
            x = j % 80 + 1
            if i == ExpectedWeaponNum-1:
                TPmatrix[offset+j, offset+400+CharacterPoolOffset] = percent_weapon(x)*0.5
            else:
                TPmatrix[offset+j, offset+400] = percent_weapon(x)*0.5
            TPmatrix[offset+j, offset+160] = percent_weapon(x)*0.5
            #在p159状态下抽卡必定成功，故一定会转移到p160状态，这里加上条件判断是为了避免覆盖前面的代码
            if j != 159:
                TPmatrix[offset+j, offset+j+1] = 1-percent_weapon(x)
        for j in range(160, 240):
            x = j % 80 + 1
            if i == ExpectedWeaponNum-1:
                TPmatrix[offset+j, offset+400+CharacterPoolOffset] = percent_weapon(x)*0.375
            else:
                TPmatrix[offset+j, offset+400] = percent_weapon(x)*0.375
            TPmatrix[offset+j, offset+320] = percent_weapon(x)*0.625
            TPmatrix[offset+j, offset+j+1] = 1-percent_weapon(x)
        for j in range(240, 320):
            x = j % 80 + 1
            if i == ExpectedWeaponNum-1:
                TPmatrix[offset+j, offset+400+CharacterPoolOffset] = percent_weapon(x)*0.5
            else:
                TPmatrix[offset+j, offset+400] = percent_weapon(x)*0.5
            TPmatrix[offset+j, offset+320] = percent_weapon(x)*0.5
            if j != 319:
                TPmatrix[offset+j, offset+j+1] = 1-percent_weapon(x)
        for j in range(320, 400):
            x = j % 80 + 1
            if i == ExpectedWeaponNum-1:
                TPmatrix[offset+j, offset+400+CharacterPoolOffset] = percent_weapon(x)
            else:
                TPmatrix[offset+j, offset+400] = percent_weapon(x)
            if j != 399:
                TPmatrix[offset+j, offset+j+1] = 1-percent_weapon(x)
    #这一部分代码生成抽角色的状态，如果要抽的角色数为0，那么就不会运行这一部分代码
    for i in range(0, ExpectedCharacterNum):
        offset = 180*i+ExpectedWeaponNum*400
        for j in range(0, 90):
            x = j % 90 + 1
            TPmatrix[offset+j, offset+180] = percent_character(x)*0.5
            TPmatrix[offset+j, offset+90] = percent_character(x)*0.5
            if j != 89:
                TPmatrix[offset+j, offset+j+1] = 1-percent_character(x)
        for j in range(90, 180):
            x = j % 90 + 1
            TPmatrix[offset+j, offset+180] = percent_character(x)
            if j != 179:
                TPmatrix[offset+j, offset+j+1] = 1-percent_character(x)
    #最后一行表示已经达成抽卡预期，所以从该状态到其他状态的概率都是0，到自身的概率为1
    TPmatrix[size-1, size-1] = 1
    #生成初始状态向量，如果抽武器，那么和武器池水位有关，否则和角色池水位有关
    initVector = np.zeros((size))
    if ExpectedWeaponNum != 0:
        if BindingNum == 0:
            if WeaponPoolGuarantee == False:
                initVector[WeaponPoolStage] = 1
            elif WeaponPoolGuarantee == True:
                initVector[WeaponPoolStage+80] = 1
        elif BindingNum == 1:
            if WeaponPoolGuarantee == False:
                initVector[WeaponPoolStage+160] = 1
            elif WeaponPoolGuarantee == True:
                initVector[WeaponPoolStage+240] = 1
        elif BindingNum == 2:
            initVector[WeaponPoolStage+320] = 1
    else:#这里是不抽武器的情况，和角色池水位有关
        initVector[CharacterPoolOffset] = 1
    #将初始状态向量和转移概率矩阵不断相乘，相乘的次数为抽数，得到预期次数后状态的概率分布
    result = initVector
    for i in range(0, IntertwinedFateNum):
        result = result@TPmatrix
    #这里取状态概率向量的最后一个元素，即已经抽到预期的角色和武器的概率
    print("当前情况下抽到",ExpectedCharacterNum,"个限定角色和",ExpectedWeaponNum,"把限定武器的概率为",(result[size-1]*100).round(2), "%")


#拥有的纠缠之缘数量
IntertwinedFateNum = 301
########################## 下为角色池部分 ##########################
#期望抽到角色数（0-7）
ExpectedCharacterNum = 3
#当前是否大保底（True/False）
CharacterPoolGuarantee = True
#角色池的水位（0-89）
CharacterPoolStage = 89
########################## 下为武器池部分 ##########################
#期望抽到武器数（0-5）
ExpectedWeaponNum = 2
#当前是否大保底（True/False）
WeaponPoolGuarantee = False
#武器池的水位（0-79）
WeaponPoolStage = 0
#命定值（0-2）
BindingNum = 0

main()

File: WishCalculator.js
import plugin from '../../lib/plugins/plugin.js'
import { segment } from 'oicq'
import common from "../../lib/common/common.js";
import { create, all } from 'mathjs'
const config = { }
const math = create(all, config)

export class WishCalculator extends plugin {
    constructor() {
        super({
        name: "祈愿概率计算",
        dsc: "祈愿概率计算",
        /** https://oicqjs.github.io/oicq/#events */
        event: "message.group",
        priority: 200,
        rule: [
            {
                reg: /^\d{1,4}，\d，\d，\d{1,2}，\d，\d，\d{1,2}，\d$/,
                fnc: "calculate",
            },
            {
                reg: /^#*(概率|期望|计算)$/,
                fnc: "help",
            }
        ]
        });
    }

    percent_character(x) {
        if (x<=73) {
            return 0.006
        } else if (x<=89) {
            return (0.006+0.06*(x-73))
        } else {
            return 1
        }
    }

    percent_weapon(x) {
        if (x<=62) {
            return 0.007
        } else if (x<=73) {
            return (0.007+0.07*(x-62))
        } else if (x<=79){
            return (0.777+0.035*(x-73))
        } else {
            return 1
        }
    }

    async calculate(e) {
        let message = e.msg
        let numbers = message.split('，')
        if (parseInt(numbers[0]) > 1200) {
            await e.reply("非法数据，抽数大于1200")
            return
        }
        if (parseInt(numbers[1]) > 7) {
            await e.reply("非法数据，预期抽到角色数大于7")
            return
        }
        if (parseInt(numbers[3]) > 89) {
            await e.reply("非法数据，角色池水位大于89")
            return
        }
        if (parseInt(numbers[4]) > 5) {
            await e.reply("非法数据，预期抽到武器数大于5")
            return
        }
        if (parseInt(numbers[6]) > 79) {
            await e.reply("非法数据，武器池水位大于79")
            return
        }
        if (parseInt(numbers[7]) > 2) {
            await e.reply("非法数据，命定值大于2")
            return
        }
        let nickname = e.sender.card
        await e.reply(`开始为${nickname}计算抽取${numbers[1]-1}命${numbers[4]}精的概率，请稍等`)
        // #拥有的纠缠之缘数量
        let IntertwinedFateNum = parseInt(numbers[0])
        // #期望抽到角色数（0-7）
        let ExpectedCharacterNum = parseInt(numbers[1])
        // #当前是否大保底（True/False）
        let CharacterPoolGuarantee = parseInt(numbers[2])
        if (CharacterPoolGuarantee == 0) {
            CharacterPoolGuarantee = false
        } else {
            CharacterPoolGuarantee = true
        }
        // #角色池的水位（0-89）
        let CharacterPoolStage = parseInt(numbers[3])
        // #期望抽到武器数（0-5）
        let ExpectedWeaponNum = parseInt(numbers[4])
        // #当前是否大保底（True/False）
        let WeaponPoolGuarantee = parseInt(numbers[5])
        if (WeaponPoolGuarantee == 0) {
            WeaponPoolGuarantee = false
        } else {
            WeaponPoolGuarantee = true
        }
        // #武器池的水位（0-79）
        let WeaponPoolStage = parseInt(numbers[6])
        // #命定值（0-2）
        let BindingNum = parseInt(numbers[7])

        let size = 180*ExpectedCharacterNum+400*ExpectedWeaponNum+1
        let TPmatrix = math.zeros((size, size))
        let CharacterPoolOffset = 0

        if (ExpectedCharacterNum != 0) {
            if (CharacterPoolGuarantee == false) {
                CharacterPoolOffset = CharacterPoolStage
            } else {
                CharacterPoolOffset = CharacterPoolStage+90
            }
        }

        for (let i = 0; i < ExpectedWeaponNum; i++) {
            let offset = 400*i
            for (let j = 0; j < 80; j++) {
                let percent = this.percent_weapon(j % 80 + 1)
                if (i == ExpectedWeaponNum-1) {
                    TPmatrix.set([offset+j, offset+400+CharacterPoolOffset],percent*0.375)
                } else {
                    TPmatrix.set([offset+j, offset+400],percent*0.375)
                }
                TPmatrix.set([offset+j, offset+160],percent*0.375)
                TPmatrix.set([offset+j, offset+240],percent*0.25)
                TPmatrix.set([offset+j, offset+j+1],1-percent)
            }
            for (let j = 80; j < 160; j++) {
                let percent = this.percent_weapon(j % 80 + 1)
                if (i == ExpectedWeaponNum-1) {
                    TPmatrix.set([offset+j, offset+400+CharacterPoolOffset],percent*0.5)
                } else {
                    TPmatrix.set([offset+j, offset+400],percent*0.5)
                }
                TPmatrix.set([offset+j, offset+160],percent*0.5)
                if (j != 159) {
                    TPmatrix.set([offset+j, offset+j+1],1-percent)
                }
            }
            for (let j = 160; j < 240; j++) {
                let percent = this.percent_weapon(j % 80 + 1)
                if (i == ExpectedWeaponNum-1) {
                    TPmatrix.set([offset+j, offset+400+CharacterPoolOffset],percent*0.375)
                } else {
                    TPmatrix.set([offset+j, offset+400],percent*0.375)
                }
                TPmatrix.set([offset+j, offset+320],percent*0.625)
                TPmatrix.set([offset+j, offset+j+1],1-percent)
            }
            for (let j = 240; j < 320; j++) {
                let percent = this.percent_weapon(j % 80 + 1)
                if (i == ExpectedWeaponNum-1) {
                    TPmatrix.set([offset+j, offset+400+CharacterPoolOffset],percent*0.5)
                } else {
                    TPmatrix.set([offset+j, offset+400],percent*0.5)
                }
                TPmatrix.set([offset+j, offset+320],percent*0.5)
                if (j != 319) {
                    TPmatrix.set([offset+j, offset+j+1],1-percent)
                }
            }
            for (let j = 320; j < 400; j++) {
                let percent = this.percent_weapon(j % 80 + 1)
                if (i == ExpectedWeaponNum-1) {
                    TPmatrix.set([offset+j, offset+400+CharacterPoolOffset],percent)
                } else {
                    TPmatrix.set([offset+j, offset+400],percent)
                }
                if (j != 399) {
                    TPmatrix.set([offset+j, offset+j+1],1-percent)
                }
            }
        }

        for (let i = 0; i < ExpectedCharacterNum; i++) {
            let offset = 180*i+ExpectedWeaponNum*400
            for (let j = 0; j < 90; j++) {
                let percent = this.percent_character(j % 90 + 1)
                TPmatrix.set([offset+j, offset+180],percent*0.5)
                TPmatrix.set([offset+j, offset+90],percent*0.5)
                if ( j != 89) {
                    TPmatrix.set([offset+j, offset+j+1],1-percent)
                }
            }
            for (let j = 90; j < 180; j++) {
                let percent = this.percent_character(j % 90 + 1)
                TPmatrix.set([offset+j, offset+180],percent)
                if ( j != 179) {
                    TPmatrix.set([offset+j, offset+j+1],1-percent)
                }
            }
        }

        TPmatrix.set([size-1, size-1],1)

        let initVector = math.zeros(size)
        if (ExpectedWeaponNum != 0) {
            if (BindingNum == 0) {
                if (WeaponPoolGuarantee == false) {
                    initVector.set([WeaponPoolStage],1)
                } else {
                    initVector.set([WeaponPoolStage+80],1)
                }
            } else if (BindingNum == 1) {
                if (WeaponPoolGuarantee == false) {
                    initVector.set([WeaponPoolStage+160],1)
                } else {
                    initVector.set([WeaponPoolStage+240],1)
                }
            } else if (BindingNum == 2) {
                initVector.set([WeaponPoolStage+320],1)
            }
        } else {
            initVector.set([CharacterPoolOffset],1)
        }

        let result = initVector
        for (let i = 0; i < IntertwinedFateNum; i++) {
            result = math.multiply(result, TPmatrix)
        }

        let msg=`\n1、拥有粉球数（${IntertwinedFateNum}）\n`+
                `2、期望抽到角色数（${ExpectedCharacterNum}）\n`+
                `3、角色池是否大保底（${CharacterPoolGuarantee}）\n`+
                `4、角色池当前水位（${CharacterPoolStage}）\n`+
                `5、期望抽到武器数（${ExpectedWeaponNum}）\n`+
                `6、武器池是否大保底（${WeaponPoolGuarantee}）\n`+
                `7、武器池当前水位（${WeaponPoolStage}）\n`+
                `8、武器池命定值（${BindingNum}）\n`+
                "在如上条件下，达成抽卡预期的概率为"
        let res_percent = `${math.round(result.get([size-1])*100,2)}%`
        await e.reply([segment.at(e.user_id),msg+res_percent])
        return
    }

    async help(e) {
        let title = "祈愿预期概率计算帮助"
        let body = "按照如下顺序提供数据：\n"+
                    "1、拥有粉球数（1-1200）\n"+
                    "2、期望抽到角色数（0-7）\n"+
                    "3、角色池是否大保底（0/1）\n"+
                    "4、角色池当前水位（0-89）\n"+
                    "5、期望抽到武器数（0-5）\n"+
                    "6、武器池是否大保底（0/1）\n"+
                    "7、武器池当前水位（0-79）\n"+
                    "8、武器池命定值（0-2）\n\n"+
                    "例如我当前拥有400抽，计算从0开始抽2命1精的概率则发送如下消息:\n\n"+
                    "400，3，0，0，1，0，0，0"
        let warning = "注意：本插件使用转移概率矩阵计算祈愿预期概率，抽数和预期UP数越多，所需的计算时间将越长。"
        let msg = await common.makeForwardMsg(e, [title, body, warning],title)
        await e.reply(msg)
        return
    }

}